File: sources/CSharp/src/Ers/SubModel/Component/ResourceComponent.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using Ers.Engine;

namespace Ers
{
    [StructLayout(LayoutKind.Sequential)]
    public struct ResourceComponent : ICoreComponent
    {
        /// <summary>
        /// The number of input channels.
        /// </summary>
        public int NumInputChannels
        {
            get => (int)ErsEngine.ERS_ResourceComponent_GetNumInputChannels(CorePtr);
        }

        /// <summary>
        /// The number of output channels.
        /// </summary>
        public int NumOutputChannels
        {
            get => (int)ErsEngine.ERS_ResourceComponent_GetNumOutputChannels(CorePtr);
        }

        /// <summary>
        /// Close all input channels.
        /// </summary>
        public void CloseAllInput() { ErsEngine.ERS_ResourceComponent_CloseAllInput(CorePtr); }

        /// <summary>
        /// Open all input channels.
        /// </summary>
        public void OpenAllInput() { ErsEngine.ERS_ResourceComponent_OpenAllInput(CorePtr); }

        /// <summary>
        /// Close all output channels.
        /// </summary>
        public void CloseAllOutput() { ErsEngine.ERS_ResourceComponent_CloseAllOutput(CorePtr); }

        /// <summary>
        /// Open all output channels.
        /// </summary>
        public void OpenAllOutput() { ErsEngine.ERS_ResourceComponent_OpenAllOutput(CorePtr); }

        /// <summary>
        /// Get an input channel by index.
        /// </summary>
        /// <param name="index">The index of the input channel</param>
        /// <returns>The entity representing the input channel</returns>
        public Entity GetInputChannel(int index) => ErsEngine.ERS_ResourceComponent_GetInputChannel(CorePtr, (nuint)index);

        /// <summary>
        /// Get an output channel by index.
        /// </summary>
        /// <param name="index">The index of the output channel</param>
        /// <returns>The entity representing the output channel</returns>
        public Entity GetOutputChannel(int index) => ErsEngine.ERS_ResourceComponent_GetOutputChannel(CorePtr, (nuint)index);

        /// <summary>
        /// Get an interable range for input channels.
        /// </summary>
        /// <param name="resourceEntity">Resource entity with a <see cref="ResourceComponent"/></param>
        /// <returns>Iterable input channel range</returns>
        public static IEnumerable<Entity> InputChannels(Entity resourceEntity) =>
            new ChannelEnumerable(resourceEntity, enumerateOutputChannels: false);

        /// <summary>
        /// Get an interable range for output channels.
        /// </summary>
        /// <param name="resourceEntity">Resource entity with a <see cref="ResourceComponent"/></param>
        /// <returns>Iterable output channel range</returns>
        public static IEnumerable<Entity> OutputChannels(Entity resourceEntity) =>
            new ChannelEnumerable(resourceEntity, enumerateOutputChannels: true);

        private readonly struct ChannelEnumerable(Entity resourceEntity, bool enumerateOutputChannels) : IEnumerable<Entity>
        {
            private readonly Entity resourceEntity        = resourceEntity;
            private readonly bool enumerateOutputChannels = enumerateOutputChannels;

            public IEnumerator<Entity> GetEnumerator() => new ChannelEnumerator(resourceEntity, enumerateOutputChannels);

            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }

        private class ChannelEnumerator : IEnumerator<Entity>
        {
            private readonly Entity resourceEntity;
            private readonly int channelCount;
            private readonly bool enumerateOutputChannels;
            private int currentIndex;

            public ChannelEnumerator(Entity resourceEntity, bool enumerateOutputChannels)
            {
                Ref<ResourceComponent> c = resourceEntity.GetComponent<ResourceComponent>();

                this.resourceEntity          = resourceEntity;
                this.channelCount            = enumerateOutputChannels ? c.Value.NumOutputChannels : c.Value.NumInputChannels;
                this.enumerateOutputChannels = enumerateOutputChannels;
                this.currentIndex            = -1;
            }

            public bool MoveNext()
            {
                if (currentIndex + 1 >= channelCount)
                {
                    currentIndex = channelCount;
                    return false;
                }

                currentIndex++;
                return true;
            }

            public void Reset() => currentIndex = -1;

            object IEnumerator.Current => Current;

            public Entity Current
            {
                get {
                    if (currentIndex < 0 || currentIndex >= channelCount)
                    {
                        throw new InvalidOperationException();
                    }

                    Ref<ResourceComponent> c = resourceEntity.GetComponent<ResourceComponent>();
                    return enumerateOutputChannels ? c.Value.GetOutputChannel(currentIndex) : c.Value.GetInputChannel(currentIndex);
                }
            }

            public void Dispose()
            {
                // Nothing to dispose
            }
        }

        public static nuint CoreTypeId() => ErsEngine.ERS_ResourceComponent_TypeId();

        /// <summary>
        /// Native pointer to the core instance.
        /// </summary>
        public IntPtr CorePtr
        {
            get {
                unsafe
                {
                    fixed(ResourceComponent* compPtr = &this)
                    {
                        return (IntPtr)compPtr;
                    }
                }
            }
        }
    }
}

File: sources/CPP/include/Ers/SubModel/Component/ResourceComponent.cpp
#include "ResourceComponent.h"
#include "Ers/Api.h"
#include "Ers/SubModel/Component/ChannelComponent.h"
#include "Ers/SubModel/SubModel.h"

namespace Ers
{
    uint32_t ResourceComponent::CoreTypeId()
    {
        return Ers::Engine::ERS_ResourceComponent_TypeId();
    }

    int ResourceComponent::GetNumInputChannels() const
    {
        return static_cast<int>(Ers::Engine::ERS_ResourceComponent_GetNumInputChannels(const_cast<ResourceComponent*>(this)));
    }

    int ResourceComponent::GetNumOutputChannels() const
    {
        return static_cast<int>(Ers::Engine::ERS_ResourceComponent_GetNumOutputChannels(const_cast<ResourceComponent*>(this)));
    }

    Entity ResourceComponent::GetInputChannel(int index) const
    {
        return Ers::Engine::ERS_ResourceComponent_GetInputChannel(const_cast<ResourceComponent*>(this), static_cast<size_t>(index));
    }

    Entity ResourceComponent::GetOutputChannel(int index) const
    {
        return Ers::Engine::ERS_ResourceComponent_GetOutputChannel(const_cast<ResourceComponent*>(this), static_cast<size_t>(index));
    }

    void ResourceComponent::CloseAllInput() const
    {
        Ers::Engine::ERS_ResourceComponent_CloseAllInput(const_cast<ResourceComponent*>(this));
    }

    void ResourceComponent::OpenAllInput() const
    {
        Ers::Engine::ERS_ResourceComponent_OpenAllInput(const_cast<ResourceComponent*>(this));
    }

    void ResourceComponent::CloseAllOuput() const
    {
        Ers::Engine::ERS_ResourceComponent_CloseAllOutput(const_cast<ResourceComponent*>(this));
    }

    void ResourceComponent::OpenAllOutput() const
    {
        Ers::Engine::ERS_ResourceComponent_OpenAllOutput(const_cast<ResourceComponent*>(this));
    }

    ResourceComponent::ChannelRange ResourceComponent::InputChannels(EntityID resourceEntity)
    {
        Ers::SubModel& subModel      = Ers::SubModel::Get();
        ResourceComponent* component = subModel.GetComponent<ResourceComponent>(resourceEntity);
        const int channelCount       = component ? component->GetNumInputChannels() : 0;
        return ChannelRange(resourceEntity, false, channelCount);
    }

    ResourceComponent::ChannelRange ResourceComponent::OutputChannels(EntityID resourceEntity)
    {
        Ers::SubModel& subModel      = Ers::SubModel::Get();
        ResourceComponent* component = subModel.GetComponent<ResourceComponent>(resourceEntity);
        const int channelCount       = component ? component->GetNumOutputChannels() : 0;
        return ChannelRange(resourceEntity, true, channelCount);
    }

    ResourceComponent::ChannelEnumerator ResourceComponent::ChannelRange::begin() const
    {
        return ChannelEnumerator(resourceEntity, 0, isOutputChannel);
    }

    ResourceComponent::ChannelEnumerator ResourceComponent::ChannelRange::end() const
    {
        return ChannelEnumerator(resourceEntity, channelCount, isOutputChannel);
    }

    EntityID ResourceComponent::ChannelEnumerator::operator*() const
    {
        Ers::SubModel& subModel      = Ers::SubModel::Get();
        ResourceComponent* component = subModel.GetComponent<ResourceComponent>(resourceEntity);

        return isOutputChannel ? component->GetOutputChannel(index) : component->GetInputChannel(index);
    }

    ResourceComponent::ChannelEnumerator& ResourceComponent::ChannelEnumerator::operator++()
    {
        ++index;
        return *this;
    }

    bool ResourceComponent::ChannelEnumerator::operator==(const ChannelEnumerator& other) const
    {
        return resourceEntity == other.resourceEntity && index == other.index && isOutputChannel == other.isOutputChannel;
    }

    bool ResourceComponent::ChannelEnumerator::operator!=(const ChannelEnumerator& other) const
    {
        return !(*this == other);
    }
} // namespace Ers
